In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [2]:
#import the csv Data file as pandas dataframe
df = pd.read_csv('data/StudentsPerformance.csv')
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [3]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

### 2.2 Dataset information

* gender: sex of students -> (Male/female)
* race/ethnicity: ethnicity of students -> (Group A, B, C, D)
* parental level of education: parents final enducation -> (Bachelor's degree, some college, master's degree, associate's degree, high school)
* lunch: having lunch before test(standard or free/reduced)
* test preperation course: complete or not compelete before test 
* math score
* reading score
* writing score

### 3.1  Data checks to perform
* check missing values
* check Duplicates
* check data type
* check the number of unique valuess of each column
* check statistics of data set
* check various categories present in different categorical column



In [4]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

There is no missing value in the dataset

In [5]:
df.duplicated().sum()

0

There are no duplicate values in the dataset

### 3.3 check data types

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


In [7]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

### 3.5 Check statistics of data set

In [8]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

* from the above description ofn numerical data, all means are very close  to each other- between 66 and 68.05;
* All standard deviations are also close -between 14.6 and 15.19;
* while there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading myet higher = 17 

### 3. 7 Exploring Data

In [17]:
columns = list(df.columns)
columns

['gender',
 'race/ethnicity',
 'parental level of education',
 'lunch',
 'test preparation course',
 'math score',
 'reading score',
 'writing score',
 'total score',
 'average']

In [9]:

ls = ['gender', 'race/ethnicity','parental level of education', 'lunch', 'test preparation course']
for i in ls:
    print(f"Categories in  '{i}': ", end=" ")
    print(df[i].unique())

Categories in  'gender':  ['female' 'male']
Categories in  'race/ethnicity':  ['group B' 'group C' 'group A' 'group D' 'group E']
Categories in  'parental level of education':  ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
Categories in  'lunch':  ['standard' 'free/reduced']
Categories in  'test preparation course':  ['none' 'completed']


In [16]:
# define numerica & categorical columns
numeric_features = [feature for feature in columns if df[feature].dtype !='o']
categorical_features = [feature for feature in columns if df[feature].dtype == 'o']

#print columns
print("We have {} numerica features: {}".format(len(numeric_features), numeric_features))
print('\n We have {} categorical features: {}'.format(len(categorical_features), categorical_features))


We have 10 numerica features: ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course', 'math score', 'reading score', 'writing score', 'total score', 'average']

 We have 0 categorical features: []


### 3.8 Adding columns for "Total Score" and "Average"

In [11]:
df['total score'] =  df['math score']+ df['reading score']+ df['writing score']
df['average'] = df['total score']/3
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  \
0                    none          72             72             74   
1               completed          69             90             88   
2                    none          90             95             93   
3                    none          47             57             44   
4                    none          76             78             75   

   total score    average  
0          218  72.666667  
1          247  82.333333  
2          278  92.666667  
3          148  49.333333  
4          229  76.333333